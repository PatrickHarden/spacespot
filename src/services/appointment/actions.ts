import {
  AppointmentAction,
  Appointment,
  EditAppointment,
  NewAppointment,
} from './types'
import constants from './constants'

const get = (): AppointmentAction => ({
  type: constants.GET,
})
const getSuccess = (payload: Array<Appointment>) => ({
  type: constants.GET_SUCCESS,
  payload,
})
const getError = (payload: string): AppointmentAction => ({
  type: constants.GET_ERROR,
  payload,
})
const update = (): AppointmentAction => ({
  type: constants.UPDATE,
})

const create = (payload: NewAppointment): AppointmentAction => ({
  type: constants.CREATE,
  payload,
})

const createSuccess = (payload: Appointment) => ({
  type: constants.CREATE_SUCCESS,
  payload,
})
const createError = (payload: string): AppointmentAction => ({
  type: constants.CREATE_ERROR,
  payload,
})

const accept = (payload: string): AppointmentAction => ({
  type: constants.ACCEPT,
  payload,
})

const acceptSuccess = (payload: Appointment) => ({
  type: constants.ACCEPT_SUCCESS,
  payload,
})

const acceptError = (payload: string): AppointmentAction => ({
  type: constants.ACCEPT_ERROR,
  payload,
})

const cancel = (payload: string): AppointmentAction => ({
  type: constants.CANCEL,
  payload,
})

const cancelSuccess = (payload: Appointment) => ({
  type: constants.CANCEL_SUCCESS,
  payload,
})

const cancelError = (payload: string): AppointmentAction => ({
  type: constants.CANCEL_ERROR,
  payload,
})

const change = (payload: EditAppointment): AppointmentAction => ({
  type: constants.CHANGE,
  payload,
})

const changeSuccess = (payload: Appointment) => ({
  type: constants.CHANGE_SUCCESS,
  payload,
})

const changeError = (payload: string): AppointmentAction => ({
  type: constants.CHANGE_ERROR,
  payload,
})
export default {
  get,
  getSuccess,
  getError,
  update,
  create,
  createSuccess,
  createError,
  accept,
  acceptSuccess,
  acceptError,
  cancel,
  cancelSuccess,
  cancelError,
  change,
  changeSuccess,
  changeError,
}
