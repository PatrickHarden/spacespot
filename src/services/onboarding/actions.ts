import types from 'services/onboarding/constants'
import history from 'browserhistory'
import { DropItem } from 'components/OnBoarding/DropZone'
import { FitoutOptionProps } from 'services/negotiation/types'
import {
  CommonPhotos1,
  CommonBrochures1,
} from 'services/space/PropertyListingschema'
import { Space } from 'services/space/types'

import {
  OnBoardingAction,
  OnBoardingState,
  AvailabilityFlex,
  AmenityOptions,
  SpaceStatus,
} from 'services/onboarding/types'
import { BuildingFormFields } from 'components/OnBoarding/BuildingDetailsForm'

const clear = () => ({ type: types.CLEAR })

const init = (payload: OnBoardingState): OnBoardingAction => ({
  type: types.INIT,
  payload,
})

const initSuccess = (payload: any): any => ({
  type: types.INIT_SUCCESS,
  payload,
})

const initError = (payload: string): any => ({
  type: types.INIT_ERROR,
  payload,
})

const goToBuilding = () => {
  history.push('/onboarding/')
  return {
    type: types.NOP,
  }
}

const goToSpace = (id: string) => {
  history.push(`/onboarding/space/${id}`)
  return {
    type: types.NOP,
  }
}

const goToPublish = () => {
  history.push('/onboarding/publish')
  return {
    type: types.NOP,
  }
}

const resetSpaces = () => ({
  type: types.RESET_SPACES,
})

const addNewSpace = () => ({
  type: types.ADD_NEW_SPACE,
})

const addNewSpaceAndGo = () => ({
  type: types.ADD_NEW_SPACE_GO,
})

const cloneSpace = (payload: string) => ({
  type: types.CLONE_SPACE,
  payload,
})

const removeSpace = (payload: string) => ({
  type: types.REMOVE_SPACE,
  payload,
})

const setIsNew = (payload: boolean) => ({
  type: types.SET_IS_NEW,
  payload,
})

const setImages = (payload: Array<DropItem>) => ({
  type: types.SET_IMAGES,
  payload,
})

const setUploadedImages = (payload: CommonPhotos1) => ({
  type: types.SET_UPLOADED_IMAGES,
  payload,
})

const setSpaceUploadedImages = (id: string, value: CommonPhotos1) => ({
  type: types.SET_SPACE_UPLOADED_IMAGES,
  payload: { id, value },
})

const setDocs = (payload: Array<DropItem>) => ({
  type: types.SET_DOCS,
  payload,
})

const setUploadedDocs = (payload: CommonBrochures1) => ({
  type: types.SET_UPLOADED_DOCS,
  payload,
})

const setSpaceUploadedFloorPlans = (id: string, value: CommonPhotos1) => ({
  type: types.SET_SPACE_UPLOADED_FLOORPLANS,
  payload: { id, value },
})

const setName = (payload: string) => ({
  type: types.SET_NAME,
  payload,
})

const setBuildingData = (payload: BuildingFormFields) => ({
  type: types.SET_BUILDING_DATA,
  payload,
})

const setDescription = (payload: string) => ({
  type: types.SET_DESCRIPTION,
  payload,
})

const setAddress = (payload: string) => ({
  type: types.SET_ADDRESS,
  payload,
})

const setCity = (payload: string) => ({
  type: types.SET_CITY,
  payload,
})

const setPostCode = (payload: string) => ({
  type: types.SET_POSTCODE,
  payload,
})

const setCountry = (payload: string) => ({
  type: types.SET_COUNTRY,
  payload,
})

const setLocation = (payload: string) => ({
  type: types.SET_LOCATION,
  payload,
})

const setLatLng = (payload: { lat: number; lng: number } | undefined) => ({
  type: types.SET_LATLNG,
  payload,
})

const setAmenities = (payload: AmenityOptions) => ({
  type: types.SET_AMENITIES,
  payload,
})

const initDefaultAmenities = () => ({
  type: types.INIT_DEFAULT_AMENITIES,
})

const setDefaultAmenities = (payload: AmenityOptions) => ({
  type: types.SET_DEFAULT_AMENITIES,
  payload,
})

const setAspects = (payload: Array<string>) => ({
  type: types.SET_ASPECTS,
  payload,
})

const setBuildingId = (payload: string) => ({
  type: types.SET_BUILDING_ID,
  payload,
})

const setIsPublishing = (payload: boolean) => ({
  type: types.SET_IS_PUBLISHING,
  payload,
})

const setError = (payload: string) => ({
  type: types.SET_ERROR,
  payload,
})

const updateBuilding = () => ({
  type: types.UPDATE_BUILDING,
})

/* Space actions */
const setSpaceType = (id: string, value: string) => ({
  type: types.SET_SPACE_TYPE,
  payload: { id, value },
})

const setSpaceName = (id: string, value: string) => ({
  type: types.SET_SPACE_NAME,
  payload: { id, value },
})

const setSpaceDescription = (id: string, value: string) => ({
  type: types.SET_SPACE_DESCRIPTION,
  payload: { id, value },
})

const setSpaceHighlights = (id: string, value: string) => ({
  type: types.SET_SPACE_HIGHLIGHTS,
  payload: { id, value },
})

const setSpaceSize = (id: string, value: number) => ({
  type: types.SET_SPACE_SIZE,
  payload: { id, value },
})

const setSpaceSizeCommon = (id: string, value: number) => ({
  type: types.SET_SPACE_SIZE_COMMON,
  payload: { id, value },
})

const setSpaceFloor = (id: string, value: number) => ({
  type: types.SET_SPACE_FLOOR,
  payload: { id, value },
})

const setSpaceMatterPort = (id: string, value: string) => ({
  type: types.SET_SPACE_MATTER_PORT,
  payload: { id, value },
})

const setSpaceFloored = (id: string, value: string) => ({
  type: types.SET_SPACE_FLOORED,
  payload: { id, value },
})

const setSpaceUse = (id: string, value: string) => ({
  type: types.SET_SPACE_USE,
  payload: { id, value },
})

const setSpaceRent = (id: string, value: number) => ({
  type: types.SET_SPACE_RENT,
  payload: { id, value },
})

const setSpaceServices = (id: string, value: number) => ({
  type: types.SET_SPACE_SERVICES,
  payload: { id, value },
})

const setSpaceServicesNotNegotiable = (id: string, value: boolean) => ({
  type: types.SET_SPACE_SERVICES_NOT_NEGOTIABLE,
  payload: { id, value },
})

const setSpaceAvailabilityFixed = (id: string, value: Date) => ({
  type: types.SET_SPACE_AVAILABILITY_FIXED,
  payload: { id, value },
})

const setSpaceMonths = (id: string, value: number) => ({
  type: types.SET_SPACE_MONTHS,
  payload: { id, value },
})

const setSpaceHotDesks = (id: string, value: AvailabilityFlex) => ({
  type: types.SET_SPACE_HOT_DESKS,
  payload: { id, value },
})

const setSpaceFixedDesks = (id: string, value: AvailabilityFlex) => ({
  type: types.SET_SPACE_FIXED_DESKS,
  payload: { id, value },
})

const setSpaceServicedOffices = (
  id: string,
  value: Array<AvailabilityFlex>,
) => ({
  type: types.SET_SPACE_SERVICED_OFFICES,
  payload: { id, value },
})

const setSpaceImages = (id: string, value: Array<DropItem>) => ({
  type: types.SET_SPACE_IMAGES,
  payload: { id, value },
})

const setSpaceFloorPlan = (id: string, value: Array<DropItem>) => ({
  type: types.SET_SPACE_FLOORPLAN,
  payload: { id, value },
})

const setSpaceFitout = (id: string, value: Array<FitoutOptionProps>) => ({
  type: types.SET_SPACE_FITOUT,
  payload: { id, value },
})

const setSpaceStatus = (id: string, value: SpaceStatus) => ({
  type: types.SET_SPACE_STATUS,
  payload: { id, value },
})

const updateSpace = () => ({
  type: types.UPDATE_SPACE,
})

const deleteSpace = (id: string) => ({
  type: types.DELETE_SPACE,
  payload: id,
})

const deleteBuilding = (id: string) => ({
  type: types.DELETE_BUILDING,
  payload: id,
})

const onboardingEditBuilding = (payload: Space) => ({
  type: types.EDIT_BUILDING,
  payload,
})

const onboardingEditSpace = (payload: Space) => ({
  type: types.EDIT_SPACE,
  payload,
})

export default {
  clear,
  init,
  initSuccess,
  initError,
  goToBuilding,
  goToSpace,
  goToPublish,
  addNewSpace,
  addNewSpaceAndGo,
  cloneSpace,
  deleteSpace,
  deleteBuilding,
  removeSpace,
  setIsNew,
  setImages,
  setUploadedImages,
  setSpaceUploadedImages,
  setDocs,
  setUploadedDocs,
  setSpaceUploadedFloorPlans,
  setBuildingData,
  setName,
  setDescription,
  setAddress,
  setCity,
  setPostCode,
  setCountry,
  setLocation,
  setLatLng,
  setAmenities,
  initDefaultAmenities,
  setDefaultAmenities,
  setAspects,
  setBuildingId,
  setIsPublishing,
  setError,
  setSpaceType,
  setSpaceName,
  setSpaceDescription,
  setSpaceHighlights,
  setSpaceSize,
  setSpaceSizeCommon,
  setSpaceFloor,
  setSpaceMatterPort,
  setSpaceFloored,
  setSpaceUse,
  setSpaceRent,
  setSpaceServices,
  setSpaceServicesNotNegotiable,
  setSpaceAvailabilityFixed,
  setSpaceMonths,
  setSpaceHotDesks,
  setSpaceFixedDesks,
  setSpaceServicedOffices,
  setSpaceImages,
  setSpaceFloorPlan,
  setSpaceFitout,
  setSpaceStatus,
  updateBuilding,
  updateSpace,
  resetSpaces,
  onboardingEditBuilding,
  onboardingEditSpace,
}
