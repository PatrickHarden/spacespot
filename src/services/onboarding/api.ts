import { get } from 'lodash'
import axios, { AxiosResponse } from 'axios'
import { SpaceDTO, BuildingDTO } from './types'
import { dataURItoBlob } from 'services/global/util'
import CONFIG from 'config'
import Analytics from 'services/analytics'

export const getBuildingId = async (
  token: string,
  country: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/autogeneratedid?idType=BUILDING&countryCode=${country}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.get(url, config)
  return get(resp, 'data')
}

export const getSpaceId = async (
  token: string,
  country: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/autogeneratedid?idType=SPACE&countryCode=${country}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.get(url, config)
  return get(resp, 'data')
}

export const putDocument = async (
  item: any,
  spaceId: string,
  token: string,
  fileCategory: 'pictures' | 'brochures' | 'floorplans',
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/${spaceId}/${fileCategory}`
  const formData = new FormData()
  formData.append('file', dataURItoBlob(item.uri), item.name)
  const resp: AxiosResponse<string> = await axios.post(url, formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
      Authorization: `Bearer ${token}`,
    },
  })
  return get(resp, 'data.url')
}

export const createBuilding = async (
  params: BuildingDTO,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/confirmation/building`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.post(url, params, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Add new building request',
    //value: userName
  })
  return get(resp, 'data')
}

export const createSpace = async (
  params: SpaceDTO,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/confirmation/space`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.post(url, params, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Add new Space request',
  })
  return get(resp, 'data')
}

export const updateBuilding = async (
  params: BuildingDTO,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/update/buildings/${params.id}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.post(url, params, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Update building request',
  })
  return get(resp, 'data')
}

export const updateSpace = async (
  id: string,
  params: SpaceDTO,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/update/spaces/${id}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.post(url, params, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Update space request',
  })
  return get(resp, 'data')
}

export const deleteSpace = async (
  id: string,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/delete/spaces/${id}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.delete(url, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Delete space request',
  })
  return get(resp, 'data')
}

export const deleteBuilding = async (
  id: string,
  token: string,
): Promise<string> => {
  const url = `${CONFIG.API_HOST}/onboarding/delete/buildings/${id}`
  const config = { headers: { Authorization: `Bearer ${token}` } }
  const resp: AxiosResponse<string> = await axios.delete(url, config)
  Analytics.event({
    category: 'Onboarding',
    action: 'Delete building request',
  })
  return get(resp, 'data')
}
