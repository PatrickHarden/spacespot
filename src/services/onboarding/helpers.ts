import { get, cloneDeep } from 'lodash'
import { Space } from 'services/space/types'
import spaceSelectors from 'services/space/selectors'
import {
  AmenityOptions,
  OnBoardingState,
  AvailabilityFlexPriceInterval,
  AvailabilityFlex,
} from 'services/onboarding/types'

export const aspects2Amenities = (
  amenities: AmenityOptions,
  aspects: Array<string>,
): AmenityOptions => {
  const newAmenities = cloneDeep(amenities) || {}
  if (!aspects) {
    return newAmenities
  }
  aspects.forEach(key => {
    if (newAmenities[key]) {
      newAmenities[key].checked = true
    } else {
      newAmenities[key] = {
        checked: true,
        desc: key,
      }
    }
  })
  return newAmenities
}

export const loadOnboardingSpace = (
  defaultAmenities: AmenityOptions | undefined,
  space: Space,
  isBuilding: boolean,
): OnBoardingState => {
  const building = spaceSelectors.translateBuilding(space, isBuilding)
  return {
    buildingId: building.buildingId,
    isNew: false,
    images: building.images || [],
    docs: [],
    name: building.name,
    description: building.description,
    address: building.address,
    city: building.city,
    postCode: building.postCode,
    country: building.country,
    location: building.location || '',
    latLng: building.latLng,
    district: building.district,
    subdistrict: building.subdistrict,
    amenities: aspects2Amenities(
      defaultAmenities || {},
      get(space, 'Common.Aspects', []),
    ),
    spaces: isBuilding ? {} : spaceSelectors.translateSpace(space),
    spaceCount: 0,
    uploadedImages: building.uploadedImages,
    uploadedDocs: building.uploadedDocs,
    isPublishing: false,
    defaultAmenities: defaultAmenities,
  }
}

const defaultDate = new Date()
defaultDate.setDate(new Date().getDate() + 1)

export const getDefaultAvailability = (): AvailabilityFlex => ({
  desks: 0,
  availableFrom: defaultDate,
  minLease: 0,
  price: 0,
  currencyCode: 'NOK',
  frequency: AvailabilityFlexPriceInterval.Monthly,
})
