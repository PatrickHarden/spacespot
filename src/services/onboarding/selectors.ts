import { get } from 'lodash'
import { SpacespotState } from '../global/types'
import {
  OnBoardingState,
  SpaceData,
  SpaceStatus,
  AmenityOptions,
} from './types'
import { TranslatePropSpace } from '../space/types'
import { getRent, getSpace, getAvailability } from './utils'

const onboardingState = (state: SpacespotState): OnBoardingState =>
  get(state, 'onboarding', {} as OnBoardingState)
const onboardingState2Space = (
  state: SpacespotState,
): Array<TranslatePropSpace> => {
  const onboarding: OnBoardingState = get(
    state,
    'onboarding',
    {} as OnBoardingState,
  )

  return Object.keys(onboarding.spaces).map((key: string) => {
    const space: SpaceData = get(onboarding, `spaces[${key}]`)
    const images = [...space.images, ...onboarding.images]
    return {
      images,
      key: space.id,
      buildingId: onboarding.buildingId,
      name: space.spaceName,
      address: onboarding.address,
      city: onboarding.city,
      postCode: onboarding.postCode,
      country: onboarding.country,
      description: space.spaceDescription,
      latLng: onboarding.latLng
        ? { lat: onboarding.latLng.lat, lng: onboarding.latLng.lng }
        : { lat: 0, lng: 0 },
      district: onboarding.district,
      subdistrict: onboarding.subdistrict,
      aspects: null,
      floor: space.spaceFloor || 0,
      type: space.type,
      rent: getRent(space).toString(),
      size: getSpace(onboarding, key),
      availability: getAvailability(space),
      isNew: space.status === SpaceStatus.New,
      hotDesks: get(space, 'hotDesks.desks', 0),
      fixedDesks: get(space, 'fixedDesks.desks', 0),
      sevicedOffices: get(space, 'servicedOffices.length', 0),
      status: get(space, 'status'),
    }
  })
}

const onboardingIsPublishing = (state: SpacespotState): boolean =>
  get(state, 'onboarding.isPublishing', false)

const defaultAmenities = (state: SpacespotState): AmenityOptions =>
  get(state, 'onboarding.defaultAmenities')

const onboardingCountry = (state: SpacespotState): string =>
  get(state, 'onboarding.country', 'NO')

export default {
  defaultAmenities,
  onboardingCountry,
  onboardingState,
  onboardingState2Space,
  onboardingIsPublishing,
}
