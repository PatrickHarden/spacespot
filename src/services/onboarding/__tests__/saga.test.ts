import { getAmenities, AmenityData } from 'services/config/api'
import { all, call, put, select } from 'redux-saga/effects'
import userSelectors from 'services/user/selectors'
import { createOnBoardingState } from 'components/OnBoarding/helpers'
import { CommonImage } from 'services/space/PropertyListingschema'
import { getRegion } from 'services/global/region'
import {
  publishOnboardingSaga,
  publishDoc,
  publishSpace,
  createResources,
  createBuildingDTO,
  getBuildingDetails,
  getSpaceDetails,
  createSpaceDTO,
  getFixedDetails,
  getFlexDetails,
  createResourcesImagesFromAPI,
  updateBuildingSaga,
  updateSpaceSaga,
  addSpaceSaga,
  initAmenitiesSaga,
} from '../saga'
import {
  getBuildingId,
  getSpaceId,
  createSpace,
  createBuilding,
  putDocument,
  updateSpace,
} from '../api'
import actions from '../actions'
import selectors from '../selectors'
import { AmenityOptions, SpaceStatus } from '../types'

type Saga = Generator<void, any, any>

describe('Onboarding', () => {
  it('saga should ', async () => {
    const images = [{ uri: '.', name: 'img1', key: 12, type: 'image/png' }]
    const docs = [{ uri: '.', name: 'doc1', key: 12, type: 'application/pdf' }]
    const token = 'dummy'
    const state = createOnBoardingState()
    state.images = images
    state.docs = docs
    const pDocs = {
      photos: createResources([], state.images, true),
      brochures: createResources([], state.docs, false),
      floorPlans: [],
    }
    const building = createBuildingDTO('id1', pDocs, state)
    const gen = publishOnboardingSaga(actions.init(state)) as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    expect(gen.next(token).value).toEqual(call(getBuildingId, token))
    expect(gen.next('id1').value).toEqual(
      all(images.map(i => call(publishDoc, i, 'id1', 'pictures'))),
    )
    expect(gen.next([]).value).toEqual(
      all(docs.map(i => call(publishDoc, i, 'id1', 'brochures'))),
    )
    expect(gen.next([]).value).toEqual(call(createBuilding, building, token))
    expect(gen.next().value).toEqual(put(actions.setIsNew(false)))
    expect(gen.next().value).toEqual(put(actions.setBuildingId('id1')))
    expect(gen.next().value).toEqual(
      all(
        Object.keys(state.spaces)
          .filter(
            key => state.spaces[key].status === SpaceStatus.ReadyForPublish,
          )
          .map(key => call(publishSpace, state, 'id1', key)),
      ),
    )
    expect(gen.next().done).toEqual(true)
  })

  it('building saga should fail', async () => {
    const state = createOnBoardingState()
    const gen = publishOnboardingSaga(actions.init(state)) as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    gen.throw(new Error('Error1'))
  })

  it('space saga', async () => {
    const token = 'dummy'
    const images = [{ uri: '.', name: 'img1', key: 12, type: 'image/png' }]
    const plans = [{ uri: '.', name: 'img1', key: 12, type: 'image/png' }]
    const state = createOnBoardingState()
    state.spaces[0].images = images
    state.spaces[0].floorPlan = plans
    const gen = publishSpace(state, 'bid1', '0') as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    expect(gen.next(token).value).toEqual(call(getSpaceId, token))
    expect(gen.next('id1').value).toEqual(
      all(images.map((i: any) => call(publishDoc, i, 'id1', 'pictures'))),
    )
    expect(gen.next([]).value).toEqual(
      all(plans.map((i: any) => call(publishDoc, i, 'id1', 'floorplans'))),
    )
    const pDocs = {
      photos: createResources([], images, true),
      brochures: [],
      floorPlans: createResources([], plans, true),
    }
    const buildingDetails = getBuildingDetails(state)
    const amenities = ['hasGym']
    const spaceDetails = getSpaceDetails(state, '0')
    const spaceDTO = createSpaceDTO(
      'bid1',
      'id1',
      pDocs,
      amenities,
      buildingDetails,
      spaceDetails,
      '',
    )
    expect(gen.next([]).value).toEqual(call(createSpace, spaceDTO, token))
    expect(gen.next().value).toEqual(
      put(actions.setSpaceStatus('0', SpaceStatus.Published)),
    )
    expect(gen.next().done).toEqual(true)
  })

  it('space saga should fail', async () => {
    const state = createOnBoardingState()
    const gen = publishSpace(state, 'id1', '0') as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    gen.throw(new Error('Error1'))
  })

  it('doc saga', async () => {
    const token = 'dummy'
    const image = { uri: '.', name: 'img1', key: 12, type: 'image/png' }
    const gen = publishDoc(image, 'id1', 'pictures') as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    expect(gen.next(token).value).toEqual(
      call(putDocument, image, 'id1', token, 'pictures'),
    )
    expect(gen.next().done).toEqual(true)
  })

  it('doc saga should fail', async () => {
    const image = { uri: '.', name: 'img1', key: 12, type: 'image/png' }
    const gen = publishDoc(image, 'id1', 'pictures') as Saga
    expect(gen.next().value).toEqual(select(userSelectors.token))
    gen.throw(new Error('Error1'))
  })

  it('helpers', async () => {
    const state = createOnBoardingState()
    expect(getFixedDetails(state, '0')).toMatchSnapshot()
    expect(getFlexDetails(state, '0')).toMatchSnapshot()
    expect(createResources([], state.images, true)).toMatchSnapshot()
    expect(createResources([], state.images, false)).toMatchSnapshot()

    const img: CommonImage = {
      'Common.ImageResources': [
        {
          'Common.Image.Width': 100,
          'Common.Image.Height': 100,
          'Common.Resource.Uri': 'https://www.spacespot.com',
        },
      ],
    }
    const items: Array<CommonImage> = [img, img, img]
    const rsrcs = createResourcesImagesFromAPI(items)
    expect(rsrcs).toMatchSnapshot()
  })

  it('update building', async () => {
    const token = 'dummy'
    const image = { uri: '.', name: 'img1', key: 12, type: 'image/png' }
    const images = [image]
    const docs = [image]
    const state = createOnBoardingState()
    state.images = images
    state.docs = docs
    state.buildingId = 'id1'
    const gen = updateBuildingSaga() as Saga
    expect(gen.next().value).toEqual(put(actions.setIsPublishing(true)))
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.next(state).value).toEqual(select(userSelectors.token))
    expect(gen.next(token).value).toEqual(
      all(images.map((i: any) => call(publishDoc, i, 'id1', 'pictures'))),
    )
    expect(gen.next([]).value).toEqual(
      all(docs.map((i: any) => call(publishDoc, i, 'id1', 'brochures'))),
    )
    expect(gen.next([]).value).toEqual(
      jasmine.objectContaining({
        type: 'CALL',
        payload: jasmine.any(Object),
      }),
    )
    expect(gen.next().value).toEqual(put(actions.setIsPublishing(false)))
    expect(gen.next().done).toEqual(true)
  })

  it('update building should fail without building', async () => {
    const gen = updateBuildingSaga() as Saga
    expect(gen.next().value).toEqual(put(actions.setIsPublishing(true)))
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.next({}).value).toEqual(select(userSelectors.token))
    expect(gen.next().value).toEqual(
      put(actions.setError((jasmine.any(String) as unknown) as string)),
    )
  })

  it('update building should fail', async () => {
    const gen = updateBuildingSaga() as Saga
    expect(gen.next().value).toEqual(put(actions.setIsPublishing(true)))
    expect(gen.throw(new Error('Error1')).value).toEqual(
      put(actions.setError('Error1')),
    )
  })

  it('update space', async () => {
    const gen = updateSpaceSaga() as Saga
    const token = 'dummy'
    const images = [{ uri: '.', name: 'img1', key: 12, type: 'image/png' }]
    const plans = [{ uri: '.', name: 'img1', key: 12, type: 'image/png' }]
    const state = createOnBoardingState()
    state.buildingId = 'bid1'
    state.spaces[0].images = images
    state.spaces[0].floorPlan = plans
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.next(state).value).toEqual(select(userSelectors.token))
    expect(gen.next(token).value).toEqual(
      all(images.map((i: any) => call(publishDoc, i, '0', 'pictures'))),
    )
    expect(gen.next([]).value).toEqual(
      all(plans.map((i: any) => call(publishDoc, i, '0', 'floorplans'))),
    )
    const pDocs = {
      photos: createResources([], images, true),
      brochures: [],
      floorPlans: createResources([], plans, true),
    }
    const buildingDetails = getBuildingDetails(state)
    const amenities = ['hasGym']
    const spaceDetails = getSpaceDetails(state, '0')
    const spaceDTO = createSpaceDTO(
      'bid1',
      '0',
      pDocs,
      amenities,
      buildingDetails,
      spaceDetails,
      '',
    )
    expect(gen.next([]).value).toEqual(call(updateSpace, '0', spaceDTO, token))
    expect(gen.next().value).toEqual(
      put(actions.setSpaceStatus('0', SpaceStatus.Published)),
    )

    expect(gen.next().done).toEqual(true)
  })

  it('update space should fail without building', async () => {
    const gen = updateSpaceSaga() as Saga
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.next().value).toEqual(
      put(actions.setError((jasmine.any(String) as unknown) as string)),
    )
  })

  it('update space should fail', async () => {
    const gen = updateSpaceSaga() as Saga
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.throw(new Error('Error1')).value).toEqual(
      put(actions.setError('Error1')),
    )
  })

  it('addSpaceSaga saga', async () => {
    const gen = addSpaceSaga() as Saga
    const state = { spaceCount: 0 }
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.next(state).value).toEqual(put(actions.addNewSpace()))
    expect(gen.next(state).value).toEqual(put(actions.goToSpace('0')))
    expect(gen.next().done).toEqual(true)
  })

  it('addSpaceSaga saga fails', async () => {
    const gen = addSpaceSaga() as Saga
    expect(gen.next().value).toEqual(select(selectors.onboardingState))
    expect(gen.throw(new Error('Error1')).value).toEqual(
      put(actions.setError('Error1')),
    )
  })

  it('initAmenities saga', async () => {
    const gen = initAmenitiesSaga() as Saga
    const lang = navigator.language
    const region = getRegion()
    const data: Array<AmenityData> = [{ name: 'test', description: 'test' }]
    const amenities: AmenityOptions = {
      test: {
        checked: false,
        desc: 'test',
      },
    }
    expect(gen.next().value).toEqual(call(getAmenities, region, lang))
    expect(gen.next(data).value).toEqual(
      put(actions.setDefaultAmenities(amenities)),
    )
    expect(gen.next().done).toEqual(true)
  })

  it('initAmenities saga error', async () => {
    const gen = initAmenitiesSaga() as Saga
    expect(gen.next().value).toEqual(
      call(getAmenities, getRegion(), navigator.language),
    )
    expect(gen.throw(new Error('Error1')).value).toEqual(
      put(actions.setError('Error1')),
    )
  })
})
