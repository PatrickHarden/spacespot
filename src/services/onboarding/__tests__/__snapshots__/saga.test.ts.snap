// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding helpers 1`] = `
Object {
  "availableFrom": "2020-01-01",
  "charges": Array [
    Object {
      "currencyCode": "NOK",
      "interval": "Monthly",
      "negotiable": true,
      "type": "Rent",
      "unit": "Whole",
      "value": 0,
    },
    Object {
      "currencyCode": "NOK",
      "interval": "Monthly",
      "negotiable": true,
      "type": "ServiceCharge",
      "unit": "Whole",
      "value": 0,
    },
  ],
  "commonArea": Object {
    "unit": "sqm",
    "value": 0,
  },
  "exclusiveSize": Object {
    "unit": "sqm",
    "value": 0,
  },
  "fitOutOptions": Array [
    Object {
      "charges": Object {
        "currencyCode": "NOK",
        "interval": "Monthly",
        "negotiable": false,
        "onApplication": true,
        "type": "PropertyAccommodation",
        "value": 1000,
      },
      "description": "FitOut1",
      "identifier": "fitout",
      "name": "FitOut1",
    },
  ],
  "floor": 0,
  "minLease": 0,
  "propertyType": undefined,
}
`;

exports[`Onboarding helpers 2`] = `
Object {
  "availableSpace": Array [],
}
`;

exports[`Onboarding helpers 3`] = `Array []`;

exports[`Onboarding helpers 4`] = `Array []`;

exports[`Onboarding helpers 5`] = `
Array [
  Object {
    "caption": undefined,
    "resource": Array [
      Object {
        "height": 100,
        "uri": "https://uat.commerciallistings.cbre.co.ukhttps://www.spacespot.com",
        "width": 100,
      },
    ],
  },
  Object {
    "caption": undefined,
    "resource": Array [
      Object {
        "height": 100,
        "uri": "https://uat.commerciallistings.cbre.co.ukhttps://www.spacespot.com",
        "width": 100,
      },
    ],
  },
  Object {
    "caption": undefined,
    "resource": Array [
      Object {
        "height": 100,
        "uri": "https://uat.commerciallistings.cbre.co.ukhttps://www.spacespot.com",
        "width": 100,
      },
    ],
  },
]
`;
